File: solutions/julia/resistor-color-trio/resistor-color-trio.jl
using MLStyle

function matchcolor(color)
    @match color begin
        "black" => 0
        "brown" => 1
        "red" => 2
        "orange" => 3
        "yellow" => 4
        "green" => 5
        "blue" => 6
        "violet" => 7
        "grey" => 8
        "white" => 9
    end
end

function label(colors)
    resistance = @match colors begin
        [tens, ones, power] => 
            (matchcolor(tens) * 10 + matchcolor(ones)) * 10^(matchcolor(power))
    end

    @match resistance begin
        r && if r % 1000 == 0 end => "$(r ÷ 1000) kiloohms"
        r => "$r ohms"
    end
end

File: solutions/elixir/resistor-color-trio/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @doc """
  Calculate the resistance value in ohm or kiloohm from resistor colors
  """

  @colors %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms}
  def label(colors) do
    colors
    |> Enum.map(fn c -> @colors[c] end)
    |> compute()
    |> (fn
          r when r > 1000 -> {r / 1000, :kiloohms}
          r -> {r, :ohms}
        end).()
  end

  defp compute([tens, ones, power]) do
    (tens * 10 + ones) * Integer.pow(10, power)
  end
end
